Sources/ComposableArchitecture/Effect.swift

```diff
@@ -1,185 +1,331 @@
+  ///
+  /// - Parameters:
+  ///   - priority: Priority of the underlying task. If `nil`, the priority will come from
+  ///     `Task.currentPriority`.
+  ///   - operation: The operation to execute.
+  ///   - catch: An error handler, invoked if the operation throws an error other than
+  ///     `CancellationError`.
+  /// - Returns: An effect wrapping the given asynchronous work.
+  public static func task(
+    priority: TaskPriority? = nil,
+    operation: @escaping @Sendable () async throws -> Output,
+    catch handler: (@Sendable (Error) async -> Output)? = nil,
+    file: StaticString = #file,
+    fileID: StaticString = #fileID,
+    line: UInt = #line
+  ) -> Self {
+    Deferred<Publishers.HandleEvents<PassthroughSubject<Output, Failure>>> {
+      let subject = PassthroughSubject<Output, Failure>()
+      let task = Task(priority: priority) { @MainActor in
+        defer { subject.send(completion: .finished) }
+        do {
+          try Task.checkCancellation()
+          let output = try await operation()
+          try Task.checkCancellation()
+          subject.send(output)
+        } catch is CancellationError {
+          return
+        } catch {
+          guard let handler = handler else {
+            #if DEBUG
+              var errorDump = ""
+              customDump(error, to: &errorDump, indent: 4)
+              runtimeWarning(
+                """
+                An 'Effect.task' returned from "%@:%d" threw an unhandled error:
 
-  /// Initializes an effect that immediately fails with the error passed in.
-  ///
-  /// - Parameter error: The error that is immediately emitted by the effect.
-  public init(error: Failure) {
-    // NB: Ideally we'd return a `Fail` publisher here, but due to a bug in iOS 13 that publisher
-    //     can crash when used with certain combinations of operators such as `.retry.catch`. The
-    //     bug was fixed in iOS 14, but to remain compatible with iOS 13 and higher we need to do
-    //     a little trickery to fail in a slightly different way.
-    self.init(
-      Deferred {
-        Future { $0(.failure(error)) }
-      }
-    )
-  }
+                %@
 
-  /// An effect that does nothing and completes immediately. Useful for situations where you must
-  /// return an effect, but you don't need to do anything.
-  public static var none: Self {
-    Empty(completeImmediately: true).eraseToEffect()
+                All non-cancellation errors must be explicitly handled via the 'catch' parameter \
+                on 'Effect.task', or via a 'do' block.
+                """,
+                [
+                  "\(fileID)",
+                  line,
+                  errorDump
+                ],
+                file: file,
+                line: line
+              )
+            #endif
+            return
+          }
+          await subject.send(handler(error))
+        }
+      }
+      return subject.handleEvents(receiveCancel: task.cancel)
```

Comment:
I have an issue with cancelling task in reactiveswift-composable-architecture in the same place, e.g. changing to `subject.handleEvents(receiveCancel: { task.cancel() })` fixed the issue.

It would be great to fix for current production version, If this's the issue.

What do you think about this?

Reply:
@larryonoff Can you explain more? I'm not sure I understand.

Are you saying that you had a problem with:

```swift
subject.handleEvents(receiveCancel: task.cancel)
```

That was fixed by opening up the closure?

```swift
subject.handleEvents(receiveCancel: { task.cancel($0) }
```

And are you saying that this applied to reactiveswift-composable-architecture, or are you saying the issue also affects this TCA branch? If you're seeing a problem with our branch, can you tell us how to reproduce? We'd be happy to incorporate the change, but we want to better understand what it's fixing.

Reply:
@stephencelis sorry for such a blurry description. I wanted to describe the issue as soon as I caught it.

Let me add some details :)

I caught the issue in `reactiveswift-composable-architecture` version of TCA. I had a problem with `subject.handleEvents(receiveCancel: task.cancel)`, which was fixed by opening the closure `subject.handleEvents(receiveCancel: { task.cancel($0) }`.

I'm aware that ReactiveSwift Effect lifetime differs from Combine Effects.

I just wanted to point on the issue, because I thought that It may be easy for you to catch the issue if it's.

I checked now TCA tests and I've found a test `testCancellingTask_Infallible`. So if it passes, everything should be good.